### PR TITLE
Add "trampoline" scheduler

### PR DESCRIPTION
**Pull request by tcw165**

### CODE REVIEW

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:30]: Incorrect description. Delete or change

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:76]: what are the benefit of using `_guard` there? Do we really need to unsubscribe subscription when worker destructed?

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:94]: Can't see if we really need shared_ptr here. composite_subscription is shared_ptr itself

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:381]: For this test we don't need to wait

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:424]: Vector of strings is enough. Don't overcomplicate test

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:515]: This two code blocks looks tooooooo clooooose. Extract some function to generate such an blocks and call twice (for thread 1 and for thread 2)

**Review by victimsnino — COMMENTED** [review]
Mostly looks good

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:63]: You can use pop_with_wait there due to anyway we MUST to wait for next schedulable

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:76]: You're right. There's no actual benefit since it's in sync with the current thread.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:94]: I see. I'll reduce one layer of classes.

File: src/tests/test_schedulers.cpp
Comment (tcw165) [reply src/tests/test_schedulers.cpp:515]: In which file would you suggest for such utility function?

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:25]: @victimsnino I found this is the easiest way to create thread-local queues but I'm not sure if it's the right way. Besides, I'm seeing compile error about duplicate symbols related to this.

File: src/samples/doxygen/trampoline.cpp
Comment (tcw165) [inline src/samples/doxygen/trampoline.cpp:31]: @victimsnino I found the implementation of `just()` with the scheduler is probably unexpected:

```
        worker.schedule([iterable=std::forward<decltype(iterable)>(iterable), 
                         subscriber=std::forward<decltype(subscriber)>(subscriber), 
                         index = size_t{0}]() mutable-> schedulers::optional_duration
        {            
            try
            {
                const auto& extracted_iterable = extract_iterable_from_packed(iterable);
                const auto  end = std::cend(extracted_iterable);
                auto        itr = std::cbegin(extracted_iterable);

                std::ranges::advance(itr, static_cast<int64_t>(index), end);

                if (itr != end)
                {
                    subscriber.on_next(utils::as_const(*itr));
                    if (std::next(itr) != end) // it was not last
                    {
                        ++index;
                        return schedulers::duration{}; // re-schedule this
                    }
                }

                subscriber.on_completed();
            }
            catch(...)
            {
                subscriber.on_error(std::current_exception());
            }
            return std::nullopt;

        });
```

So there'd be only one schedulable that emits the elements one by one via `on_next()`. And that will lead to the result of `-{3,1}-{3,2}-{3,3}-|`.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:25]: > I'm seeing compile error about duplicate symbols related to this

Yes, because you've created global symbols and each compilation unit which includes this would try to create this one. Place them inside scheduler body

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [reply src/tests/test_schedulers.cpp:505]: You can extract it into same file. Or even create lambda inside a test, it doesn't matter

File: src/samples/doxygen/trampoline.cpp
Comment (victimsnino) [reply src/samples/doxygen/trampoline.cpp:31]: Probably i didn't catch your point. Can't see if anything is wrong? We need to schedule each item emission and send only one value per scheduling. Expected result would be something like this
```
> inner observable send 1
> outer observable send 1
{1,1}
> inner observable send 2
{1,2}
>outer observable send 2
{2,2}
> inner observable send 3
{2, 3}
> outer observable send 3
{3,3}
```

File: src/rpp/rpp/schedulers/details/queue_worker_state.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/details/queue_worker_state.hpp:77]: Don't overcomplicate and avoid trying to do anything as noexcept =) This function is pretty simple and can be inlined by compiler without any tips

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:32]: You need move instead of forward there. `&&` is forwarding **ONLY** when you have template argument for this one FOR function.
```cpp
template <typename F>
struct scope_guard {
    template<typename U>
    explicit scope_guard(U&& f) : m_f(std::forward<U>(f)) {}
```
^^^^^ is forwarding reference due to it is deduced FROM argument of function (constrctor). But in your case you class specialized with template and then you accept it by **rvalue-reference**

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:72]: stop token used to unblock wait in case of someone requested stop. Add it to `m_sub` to request stop in case of unsubscribe

File: src/samples/doxygen/trampoline.cpp
Comment (victimsnino) [inline src/samples/doxygen/trampoline.cpp:12]: I think we don't need it because it is overcomplication. Change to subscribe_on

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:366]: ???? Can't see why do you need it

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:371]: std::optional would be enough =)

File: src/samples/doxygen/trampoline.cpp
Comment (tcw165) [reply src/samples/doxygen/trampoline.cpp:31]: Oh I meant that I have this code and this output:

Code:
```
    rpp::source::just(rpp::schedulers::trampoline{}, 1, 2, 3)
        .with_latest_from(rpp::source::just(rpp::schedulers::trampoline{}, 4, 5, 6))
        .subscribe(
            [](const auto &v) { std::cout << v << "-"; },
            [](const std::exception_ptr &error) {},
            []() { std::cout << "|" << std::endl; });```

Output

```
-{1,6}-{2,6}-{3,6}-|
```

And the reason for that is because we only have one master schedulable that handles all the `on_next()`. Alternatively, we may consider creating new schedulable for each `on_next()`?

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:32]: You're right. Thanks for catching this :)

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:72]: Yup, I have the same question too. I'll add that later.

File: src/samples/doxygen/trampoline.cpp
Comment (victimsnino) [reply src/samples/doxygen/trampoline.cpp]: In this case you need subscribe_on. I will fix after merging your PR due to it is required extra code in subscribe method

Comment (tcw165) [reply src/samples/doxygen/trampoline.cpp]: Please let me know about the PR after this is merged.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:93]: you don't need it as shared_ptr. You can create it during `defer_at` function. Add callback to subscription and remove/unsubscribe it after while loop to remove local variable from m_sub state

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:412]: Same as above -> use optional

File: src/tests/test_schedulers.cpp
Comment (victimsnino) [inline src/tests/test_schedulers.cpp:540]: Can't see any benefit of using thread there comparing to just calling unsubscribe in place of 
```cpp
// Block return so that async stopper could prevent the inner schedulable from running.
std::this_thread::sleep_for(std::chrono::seconds{3});
```

Comment (victimsnino) [reply src/tests/test_schedulers.cpp:540]: Threads and sleeps add complexity to test, but doesn't provide any benefit and just adds race-condition.

Comment (tcw165) [reply src/tests/test_schedulers.cpp:540]: You're right! I've been having this concern for long time.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:95]: I used a shared pointer for shared ownerships. Do you mean something like below in `defer_at()`?

```
std::stop_source stop;
auto stop_token = stop.get_token();

m_sub.add([stop = std::move(stop)] { // move stop since subscription callback is the sole owner.
  stop.request_stop();
});

while (...) {
  if (queue.pop_wait(..., stop_token) ...
}
```

Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:95]: I feel the `stop_source` has the same lifetime to `m_sub` and thus I lift it as a class member.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (tcw165) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp]: Discussed offline: we use a static class function to house static `thread_local` variables.

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:30]: Use forward ONLY when template parameter is forwarding reference. In your case it is  rvalue
```suggestion
    explicit scope_guard(F&& f) : m_f(std::move<F>(f)) {}
```

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:57]: It would be more understandable =)
```suggestion
            with_thread_local_schedulable_queue([&](auto& queue, bool& queue_in_use)
```

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:75]: looks like you need to mark it as mutable to be compiled

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [inline src/rpp/rpp/schedulers/trampoline_scheduler.hpp:110]: std::function is complex and heavy type. You can use auto there or something like `const std::invocable<rpp::schedulers::details::queue_worker_state&, bool&> auto&`

File: src/rpp/rpp/schedulers/trampoline_scheduler.hpp
Comment (victimsnino) [reply src/rpp/rpp/schedulers/trampoline_scheduler.hpp:75]:
```suggestion
                m_sub.add([stop = std::move(stop)]() mutable
```

**Review by victimsnino — APPROVED** [review]

**Pull request merged**
